### PR TITLE
immutable-keyword-not-supported

### LINKED ISSUE #265: Immutable keyword not supported
Declaring an immutable storage variable:
```
uint256 immutable testVar;
```

Results in:
```
[error] contracts/Test.sol: ParserError: extraneous input 'testVar' expecting {';', '='} (136:20)
```

https://solidity.readthedocs.io/en/v0.6.5/miscellaneous.html#reserved-keywords

### PR DESCRIPTION
**Pull request by mattiaerre**
# Description

closes #265 

- add test case

- update @solidity-parser/parser

- update `VariableDeclaration `

## Notes

the test, before the parser's update, was throwing the following error:

```bash
 FAIL  tests/C/jsfmt.spec.js
  ● Test suite failed to run

    ParserError: extraneous input 'decimals' expecting {';', '='} (7:19)

      14 |
      15 | function parse(text, parsers, options) {
    > 16 |   const parsed = parser.parse(text, { loc: true, range: true });
         |                         ^
      17 |   parsed.comments = extract(text);
      18 |
      19 |   parser.visit(parsed, {

      at Object.parse (node_modules/@solidity-parser/parser/dist/index.js:79:11)
      at Object.parse (src/parser.js:16:25)
      at Object.parse (node_modules/prettier/index.js:11379:19)
      at coreFormat (node_modules/prettier/index.js:14795:25)
      at format (node_modules/prettier/index.js:15031:75)
      at formatWithCursor (node_modules/prettier/index.js:15047:12)
      at Object.args [as formatWithCursor] (node_modules/prettier/index.js:50695:12)
      at prettyprint (tests_config/run_spec.js:83:27)
      at fs.readdirSync.forEach.filename (tests_config/run_spec.js:45:22)
          at Array.forEach (<anonymous>)
```

### CODE REVIEW

File: tests/C/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+pragma solidity >0.6.4 <0.7.0;
+
+
+contract C {
+    uint256 constant X = 32**22 + 8;
+    string constant TEXT = "abc";
+    bytes32 constant MY_HASH = keccak256("abc");
+    uint256 decimals;
+    uint256 maxBalance;
```
Comment (Janther) [inline tests/C/__snapshots__/jsfmt.spec.js.snap:33]: This is solving only the parser error, `prettier-solidity` is not printing the immutable key

Comment (mattiaerre) [reply tests/C/__snapshots__/jsfmt.spec.js.snap:33]: you are right, let me mark this PR still as WIP, any suggestion?

Comment (Janther) [reply tests/C/__snapshots__/jsfmt.spec.js.snap:33]: As far as I can see, the parser isn't giving us any information right now.
there is an issue open https://github.com/solidity-parser/parser/issues/17
For the moment we wait.
 @fvictorio any comments?

Comment (fvictorio) [reply tests/C/__snapshots__/jsfmt.spec.js.snap:33]: Ugh, yeah, we'll need to add support for that in the parser. I'll work on that asap.

Comment (fvictorio) [reply tests/C/__snapshots__/jsfmt.spec.js.snap:33]: See https://github.com/solidity-parser/parser/pull/18

**Comment by fvictorio** [comment]
`@solidity-parser/parser@0.6.1` should support this now. The variable inside `StateVariableDeclaration` has a `isImmutable` field now.

**Comment by mattiaerre** [comment]
we get this error now:

```bash
FAIL tests/C/jsfmt.spec.js
  ● /home/travis/build/prettier-solidity/prettier-plugin-solidity/tests/C/C.sol parse
    expect(received).toEqual(expected) // deep equality
    - Expected  - 3
    + Received  + 3
    @@ -154,11 +154,11 @@
                "type": "StateVariableDeclaration",
                "variables": Array [
                  Object {
                    "expression": null,
                    "isDeclaredConst": false,
    -               "isImmutable": false,
    +               "isImmutable": [secure],
                    "isIndexed": false,
                    "isStateVar": [secure],
                    "name": "decimals",
                    "override": null,
                    "type": "VariableDeclaration",
    @@ -175,11 +175,11 @@
                "type": "StateVariableDeclaration",
                "variables": Array [
                  Object {
                    "expression": null,
                    "isDeclaredConst": false,
    -               "isImmutable": false,
    +               "isImmutable": [secure],
                    "isIndexed": false,
                    "isStateVar": [secure],
                    "name": "maxBalance",
                    "override": null,
                    "type": "VariableDeclaration",
    @@ -210,11 +210,11 @@
                      },
                      "memberName": "sender",
                      "type": "MemberAccess",
                    },
                    "isDeclaredConst": false,
    -               "isImmutable": false,
    +               "isImmutable": [secure],
                    "isStateVar": [secure],
                    "name": "owner",
                    "override": null,
                    "type": "VariableDeclaration",
      66 |           expect(ppastMassaged).toBeDefined();
      67 |           if (!astMassaged.errors || astMassaged.errors.length === 0) {
    > 68 |             expect(astMassaged).toEqual(ppastMassaged);
         |                                 ^
      69 |           }
      70 |         });
      71 |       }
      at Object.<anonymous> (tests_config/run_spec.js:68:33)
```

// cc @fvictorio @Janther

**Comment by mattiaerre** [comment]
> we get this error now:
> 
> ```shell
> FAIL tests/C/jsfmt.spec.js
>   ● /home/travis/build/prettier-solidity/prettier-plugin-solidity/tests/C/C.sol parse
>     expect(received).toEqual(expected) // deep equality
>     - Expected  - 3
>     + Received  + 3
>     @@ -154,11 +154,11 @@
>                 "type": "StateVariableDeclaration",
>                 "variables": Array [
>                   Object {
>                     "expression": null,
>                     "isDeclaredConst": false,
>     -               "isImmutable": false,
>     +               "isImmutable": [secure],
>                     "isIndexed": false,
>                     "isStateVar": [secure],
>                     "name": "decimals",
>                     "override": null,
>                     "type": "VariableDeclaration",
>     @@ -175,11 +175,11 @@
>                 "type": "StateVariableDeclaration",
>                 "variables": Array [
>                   Object {
>                     "expression": null,
>                     "isDeclaredConst": false,
>     -               "isImmutable": false,
>     +               "isImmutable": [secure],
>                     "isIndexed": false,
>                     "isStateVar": [secure],
>                     "name": "maxBalance",
>                     "override": null,
>                     "type": "VariableDeclaration",
>     @@ -210,11 +210,11 @@
>                       },
>                       "memberName": "sender",
>                       "type": "MemberAccess",
>                     },
>                     "isDeclaredConst": false,
>     -               "isImmutable": false,
>     +               "isImmutable": [secure],
>                     "isStateVar": [secure],
>                     "name": "owner",
>                     "override": null,
>                     "type": "VariableDeclaration",
>       66 |           expect(ppastMassaged).toBeDefined();
>       67 |           if (!astMassaged.errors || astMassaged.errors.length === 0) {
>     > 68 |             expect(astMassaged).toEqual(ppastMassaged);
>          |                                 ^
>       69 |           }
>       70 |         });
>       71 |       }
>       at Object.<anonymous> (tests_config/run_spec.js:68:33)
> ```
> 
> // cc @fvictorio @Janther

this has been fixed in the last commit 🤞

File: src/nodes/VariableDeclaration.js [hunk 2]
```diff
@@ -29,6 +31,7 @@ const VariableDeclaration = {
           visibility(node),
           constantKeyword(node),
           storageLocation(node),
+          immutable(node),
```
Comment (mattiaerre) [inline src/nodes/VariableDeclaration.js:34]: qq: is this the right position for the immutable keyword btw? // cc @fvictorio @Janther

Comment (fvictorio) [reply src/nodes/VariableDeclaration.js:34]: I think it's OK. An immutable variable cannot also be constant, and storage location doesn't make sense for state variables (which are the only ones that can be immutable), so for all effects this is just putting it after the visibility node, which makes sense.

**Comment by Janther** [comment]
I'm happy with this, but could you rename the test to immutable or variable declarations?
`C` is not very descriptive.

**Comment by mattiaerre** [comment]
> I'm happy with this, but could you rename the test to immutable or variable declarations?
> `C` is not very descriptive.

oh yeah that's the name of the contract I copied from the solidity website. I'll rename it

**Comment by mattiaerre** [comment]
and done @Janther

**Pull request merged**